File: app/src/main/java/com/ddiehl/android/reversi/game/Board.java
package com.ddiehl.android.reversi.game;


import com.ddiehl.android.reversi.exceptions.IllegalMoveException;

import java.util.Iterator;
import java.util.NoSuchElementException;

import rx.Observable;
import rx.functions.Func0;

public class Board {

    private final BoardSpace[][] spaces;
    private final int columns;
    private final int rows;

    private final byte[][] MOVE_DIRECTIONS = new byte[][] {
            {0,  -1}, // Down
            {1,  0}, // Right
            {-1, 0}, // Left
            {0,  1}, // Up
            {-1, -1}, // Down-Left
            {1,  -1}, // Down-Right
            {-1, 1}, // Top-Left
            {1,  1} // Top-Right
    };

    public Board(int rows, int columns) {
        this.columns = columns;
        this.rows = rows;
        spaces = new BoardSpace[rows][columns];
        reset();
    }

    public Board(int rows, int cols, byte[] in) {
        this(rows, cols);

        int index = 0;
        for (int y = 0; y < height(); y++) {
            for (int x = 0; x < width(); x++) {
                byte c = in[index++];

                switch (c) {
                    case 0:
                        break;
                    case 1:
                        getSpaceAt(x, y).setColor(ReversiColor.Light);
                        break;
                    case 2:
                        getSpaceAt(x, y).setColor(ReversiColor.Dark);
                }
            }
        }
    }

    public Board(int rows, int cols, String in) {
        this(rows, cols);

        int index = 0;
        for (int y = 0; y < height(); y++) {
            for (int x = 0; x < width(); x++) {
                char c = in.charAt(index++);

                switch (c) {
                    case '0':
                        break;
                    case '1':
                        getSpaceAt(x, y).setColor(ReversiColor.Light);
                        break;
                    case '2':
                        getSpaceAt(x, y).setColor(ReversiColor.Dark);
                }
            }
        }
    }

    public Board copy() {
        Board copy = new Board(rows, columns);
        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < columns; x++) {
                copy.spaces[y][x] = spaces[y][x].copy();
            }
        }
        return copy;
    }

    public void reset() {
        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < columns; x++) {
                spaces[y][x] = new BoardSpace(x, y);
            }
        }
        spaces[3][3].setColor(ReversiColor.Light);
        spaces[3][4].setColor(ReversiColor.Dark);
        spaces[4][4].setColor(ReversiColor.Light);
        spaces[4][3].setColor(ReversiColor.Dark);
    }

    public boolean hasMove(ReversiColor c) {
        BoardIterator i = iterator();
        while (i.hasNext()) {
            BoardSpace s = i.next();
            if (s.isOwned())
                continue;
            for (byte[] move : MOVE_DIRECTIONS) {
                int value = moveValueInDirection(s, move[0], move[1], c);
                if (value != 0) {
                    return true;
                }
            }
        }
        return false;
    }

    BoardSpace getSpaceAt(int x, int y) {
        if (x >= 0 && x < columns && y >= 0 && y < rows)
            return spaces[y][x];

        return null;
    }

    public Observable<Boolean> requestClaimSpace(final int x, final int y, final ReversiColor color) {
        return Observable.defer(new Func0<Observable<Boolean>>() {
            @Override
            public Observable<Boolean> call() {
                BoardSpace space = spaces[x][y];

                // If space is already claimed, return an error
                if (space.isOwned()) {
                    return Observable.error(new IllegalMoveException("space is already owned"));
                }

                int captured = spacesCapturedWithMove(space, color);
                if (captured <= 0) {
                    return Observable.error(new IllegalMoveException("move value is <= 0: " + captured));
                }

                commitPiece(space, color);
                return Observable.just(true);
            }
        });
    }

    public void commitPiece(BoardSpace space, ReversiColor playerColor) {
        for (byte[] move : MOVE_DIRECTIONS) {
            if (moveValueInDirection(space, move[0], move[1], playerColor) != 0) {
                flipInDirection(space, move[0], move[1], playerColor);
            }
        }
    }

    public int spacesCapturedWithMove(BoardSpace s, ReversiColor playerColor) {
        int moveVal = 0;
        for (byte[] move : MOVE_DIRECTIONS)
            moveVal += moveValueInDirection(s, move[0], move[1], playerColor);
        return moveVal;
    }

    private int moveValueInDirection(BoardSpace s, int dx, int dy, ReversiColor playerColor) {
        if (s.x()+dx < 0 || s.x()+dx >= columns || s.y()+dy < 0 || s.y()+dy >= rows)
            return 0;

        int moveVal = 0;
        ReversiColor opponentColor = (playerColor == ReversiColor.Dark) ? ReversiColor.Light : ReversiColor.Dark;
        BoardSpace firstPiece = getSpaceAt(s.x() + dx, s.y() + dy);

        if (firstPiece != null && firstPiece.getColor() == opponentColor) {
            int cx = s.x()+dx;
            int cy = s.y()+dy;
            while (getSpaceAt(cx, cy) != null && getSpaceAt(cx, cy).getColor() == opponentColor) {
                moveVal++;
                cx += dx;
                cy += dy;
            }
            if (getSpaceAt(cx, cy) != null && getSpaceAt(cx, cy).getColor() == playerColor) {
                return moveVal;
            }
        }

        return 0;
    }

    private void flipInDirection(BoardSpace s, int dx, int dy, ReversiColor playerColor) {
        s.setColor(playerColor);
        int cx = s.x() + dx;
        int cy = s.y() + dy;

        while (getSpaceAt(cx, cy).getColor() != playerColor) {
            getSpaceAt(cx, cy).flipColor();
            cx += dx;
            cy += dy;
        }
    }

    public int getNumSpacesForColor(ReversiColor c) {
        int count = 0;
        BoardIterator i = iterator();
        while (i.hasNext()) {
            BoardSpace s = i.next();
            if (s.isOwned() && s.getColor() == c)
                count++;
        }
        return count;
    }

    public int getNumberOfEmptySpaces() {
        int count = 0;
        BoardIterator i = iterator();
        while (i.hasNext()) {
            BoardSpace s = i.next();
            if (!s.isOwned())
                count++;
        }
        return count;
    }

    public int width() {
        return columns;
    }

    public int height() {
        return rows;
    }

    public byte[] serialize() {
        byte[] out = new byte[64];
        int index = 0;
        BoardIterator i = iterator();
        while (i.hasNext()) {
            BoardSpace s = i.next();
            if (!s.isOwned())
                out[index++] = 0;
            else if (s.getColor() == ReversiColor.Light)
                out[index++] = 1;
            else
                out[index++] = 2;
        }
        return out;
    }

    public byte getSpaceNumber(BoardSpace s) {
        return (byte) (s.y() * 8 + s.x() + 1);
    }

    public BoardSpace getBoardSpaceFromNum(int n) {
        n -= 1;
        return getSpaceAt(n % 8, n / 8);
    }

    public BoardIterator iterator() {
        return new BoardIterator(this);
    }

    public BoardSpace spaceAt(int row, int col) {
        return spaces[row][col];
    }

    static class BoardIterator implements Iterator<BoardSpace> {
        private Board mBoard;
        private int x = 0;
        private int y = 0;

        public BoardIterator(Board b) {
            mBoard = b;
        }

        @Override
        public boolean hasNext() {
            return y != mBoard.height();
        }

        @Override
        public BoardSpace next() {
            if (!hasNext()) throw new NoSuchElementException();

            BoardSpace s = mBoard.getSpaceAt(x, y);
            if (++x == mBoard.width()) {
                y++; x = 0;
            }

            return s;
        }
    }


    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("\n");
        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < columns; x++) {
                BoardSpace s = getSpaceAt(x, y);
                if (!s.isOwned())
                    sb.append("0 ");
                else if (s.getColor() == ReversiColor.Light)
                    sb.append("1 ");
                else
                    sb.append("2 ");
            }
            if (y != spaces.length-1)
                sb.append("\n");
        }
        return sb.toString();
    }
}

File: app/src/main/java/com/ddiehl/android/reversi/game/Board.kt
package com.ddiehl.android.reversi.game


import com.ddiehl.android.reversi.IllegalMoveException
import rx.Observable
import rx.functions.Func0
import java.util.*

class Board(val height: Int, val width: Int) {

    companion object {
        private val MOVE_DIRECTIONS = arrayOf(
                byteArrayOf(0, -1), // Down
                byteArrayOf(1, 0), // Right
                byteArrayOf(-1, 0), // Left
                byteArrayOf(0, 1), // Up
                byteArrayOf(-1, -1), // Down-Left
                byteArrayOf(1, -1), // Down-Right
                byteArrayOf(-1, 1), // Top-Left
                byteArrayOf(1, 1) // Top-Right
        )
    }

    private val spaces: Array<Array<BoardSpace>> =
            Array(height) { row ->
                Array(width) { col ->
                    BoardSpace(row, col)
                }
            }

    fun reset() {
        spaces[3][3].color = ReversiColor.Light
        spaces[3][4].color = ReversiColor.Dark
        spaces[4][4].color = ReversiColor.Light
        spaces[4][3].color = ReversiColor.Dark
    }

    constructor(rows: Int, cols: Int, saved: ByteArray) : this(rows, cols) {

        var index = 0
        for (y in 0..height - 1) {
            for (x in 0..width - 1) {
                val c = saved[index++]

                when (c) {
                    0.toByte() -> { }
                    1.toByte() -> getSpaceAt(x, y).color = ReversiColor.Light
                    2.toByte() -> getSpaceAt(x, y).color = ReversiColor.Dark
                }
            }
        }
    }

    constructor(rows: Int, cols: Int, saved: String) : this(rows, cols) {

        var index = 0
        for (y in 0..height - 1) {
            for (x in 0..width - 1) {
                val c = saved[index++]

                when (c) {
                    '0' -> { }
                    '1' -> getSpaceAt(x, y).color = ReversiColor.Light
                    '2' -> getSpaceAt(x, y).color = ReversiColor.Dark
                }
            }
        }
    }

    fun copy(): Board {
        val copy = Board(height, width)
        for (y in 0..height - 1) {
            for (x in 0..width - 1) {
                copy.spaces[y][x] = spaces[y][x].copy()
            }
        }
        return copy
    }

    fun hasMove(c: ReversiColor): Boolean {
        val iterator = iterator()
        while (iterator.hasNext()) {
            val space = iterator.next()
            if (!space.isOwned) {
                MOVE_DIRECTIONS
                        .map { moveValueInDirection(space, it[0].toInt(), it[1].toInt(), c) }
                        .filter { value -> value != 0 }
                        .first { return true }
            }
        }
        return false
    }

    fun getSpaceAt(x: Int, y: Int): BoardSpace =
            if (x >= 0 && x < width && y >= 0 && y < height) {
                spaces[y][x]
            } else {
                throw NullPointerException("Space @ $x $y is null")
            }

    fun requestClaimSpace(row: Int, col: Int, color: ReversiColor): Observable<Boolean> {
        return Observable.defer(Func0 {
            val space = spaces[row][col]

            // If space is already claimed, return an error
            if (space.isOwned) {
                return@Func0 Observable.error<Boolean>(IllegalMoveException("space is already owned"))
            }

            val captured = spacesCapturedWithMove(space, color)
            if (captured <= 0) {
                return@Func0 Observable.error<Boolean>(IllegalMoveException("move value is <= 0: " + captured))
            }

            commitPiece(space, color)
            Observable.just(true)
        })
    }

    fun commitPiece(space: BoardSpace, playerColor: ReversiColor) {
        MOVE_DIRECTIONS
                .filter { moveValueInDirection(space, it[0].toInt(), it[1].toInt(), playerColor) != 0 }
                .forEach { flipInDirection(space, it[0].toInt(), it[1].toInt(), playerColor) }
    }

    fun spacesCapturedWithMove(s: BoardSpace, playerColor: ReversiColor): Int =
        MOVE_DIRECTIONS.sumBy {
            moveValueInDirection(s, it[0].toInt(), it[1].toInt(), playerColor)
        }

    private fun moveValueInDirection(s: BoardSpace, dx: Int, dy: Int, playerColor: ReversiColor): Int {
        if (s.x() + dx < 0 || s.x() + dx >= width || s.y() + dy < 0 || s.y() + dy >= height)
            return 0

        var moveVal = 0
        val opponentColor = if (playerColor == ReversiColor.Dark) ReversiColor.Light else ReversiColor.Dark
        val firstPiece = getSpaceAt(s.x() + dx, s.y() + dy)

        if (firstPiece.color == opponentColor) {
            var cx = s.x() + dx
            var cy = s.y() + dy
            while (getSpaceAt(cx, cy).color == opponentColor) {
                moveVal++
                cx += dx
                cy += dy
            }
            if (getSpaceAt(cx, cy).color == playerColor) {
                return moveVal
            }
        }

        return 0
    }

    private fun flipInDirection(space: BoardSpace, dx: Int, dy: Int, playerColor: ReversiColor) {
        space.color = playerColor
        var cx = space.x() + dx
        var cy = space.y() + dy

        while (getSpaceAt(cx, cy).color != playerColor) {
            getSpaceAt(cx, cy).flipColor()
            cx += dx
            cy += dy
        }
    }

    fun getNumSpacesForColor(c: ReversiColor): Int {
        var count = 0
        val i = iterator()
        while (i.hasNext()) {
            val s = i.next()
            if (s.isOwned && s.color == c)
                count++
        }
        return count
    }

    val numberOfEmptySpaces: Int
        get() {
            var count = 0
            val i = iterator()
            while (i.hasNext()) {
                val s = i.next()
                if (!s.isOwned)
                    count++
            }
            return count
        }

    fun serialize(): ByteArray {
        val out = ByteArray(64)
        var index = 0
        val i = iterator()
        while (i.hasNext()) {
            val s = i.next()
            if (!s.isOwned)
                out[index++] = 0
            else if (s.color == ReversiColor.Light)
                out[index++] = 1
            else
                out[index++] = 2
        }
        return out
    }

    fun getSpaceNumber(s: BoardSpace): Byte {
        return (s.y() * 8 + s.x() + 1).toByte()
    }

    fun getBoardSpaceFromNum(n: Int): BoardSpace? {
        var n2 = n
        n2 -= 1
        return getSpaceAt(n % 8, n / 8)
    }

    fun iterator(): BoardIterator {
        return BoardIterator(this)
    }

    fun spaceAt(row: Int, col: Int): BoardSpace {
        return spaces[row][col]
    }

    class BoardIterator(private val mBoard: Board) : Iterator<BoardSpace> {
        private var x = 0
        private var y = 0

        override fun hasNext(): Boolean {
            return y != mBoard.height
        }

        override fun next(): BoardSpace {
            if (!hasNext()) throw NoSuchElementException()

            val s = mBoard.getSpaceAt(x, y)
            if (++x == mBoard.width) {
                y++
                x = 0
            }

            return s
        }
    }

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("\n")
        for (row in 0..height - 1) {
            (0..width - 1)
                    .map { col -> getSpaceAt(col, row) }
                    .forEach {
                        if (!it.isOwned)
                            sb.append("0 ")
                        else if (it.color == ReversiColor.Light)
                            sb.append("1 ")
                        else
                            sb.append("2 ")
                    }
            if (row != spaces.size - 1)
                sb.append("\n")
        }
        return sb.toString()
    }
}
